source/logrepl/handler.go

```diff
@@ -33,22 +35,75 @@ import (
 type CDCHandler struct {
 	tableKeys   map[string]string
 	relationSet *internal.RelationSet
-	out         chan<- opencdc.Record
-	lastTXLSN   pglogrepl.LSN
 
+	// batchSize is the largest number of records this handler will send at once.
+	batchSize     int
+	flushInterval time.Duration
+
+	// recordBatch holds the batch that is currently being built.
+	recordBatch     []opencdc.Record
+	recordBatchLock sync.Mutex
+
+	// out is a sending channel with batches of records.
+	out            chan<- []opencdc.Record
+	lastTXLSN      pglogrepl.LSN
 	withAvroSchema bool
 	keySchemas     map[string]cschema.Schema
 	payloadSchemas map[string]cschema.Schema
 }
 
-func NewCDCHandler(rs *internal.RelationSet, tableKeys map[string]string, out chan<- opencdc.Record, withAvroSchema bool) *CDCHandler {
-	return &CDCHandler{
+func NewCDCHandler(ctx context.Context, rs *internal.RelationSet, tableKeys map[string]string, out chan<- []opencdc.Record, withAvroSchema bool, batchSize int, flushInterval time.Duration) *CDCHandler {
+	h := &CDCHandler{
 		tableKeys:      tableKeys,
 		relationSet:    rs,
+		recordBatch:    make([]opencdc.Record, 0, batchSize),
 		out:            out,
 		withAvroSchema: withAvroSchema,
 		keySchemas:     make(map[string]cschema.Schema),
 		payloadSchemas: make(map[string]cschema.Schema),
+		batchSize:      batchSize,
+		flushInterval:  flushInterval,
+	}
+
+	go h.scheduleFlushing(ctx)
+
+	return h
+}
+
+func (h *CDCHandler) scheduleFlushing(ctx context.Context) {
+	ticker := time.NewTicker(h.flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
```

Comment:
This isn't quite correct, since it will flush the records no matter when the last flush happened (which can be less that `flushInterval` ago, if the batch was flushed because it reached `batchSize`). I'll change this.